### ISSUE TITLE
[Bug] ld.lld: error: duplicate symbol: columns when building r.mapcalc

### ISSUE THREAD

**Issue by attilaolah** [OSGeo/grass#1327]
**Describe the bug**

When building GRASS 7.8.5 using LLVM+Clang 11.0.1, I get the following error:

```
ld.lld: error: duplicate symbol: columns
>>> defined at map.c
>>>            OBJ.x86_64-pc-linux-gnu/map.o:(columns)
>>> defined at calc.c
>>>            calc.o:(.bss+0x4) in archive /root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/lib/libgrass_calc.7.8.a
```

The full compile + link line is as follows:

```
: && /./root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/external/llvm/bin/clang -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/lib -B/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/external/llvm/bin -fuse-ld=/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/external/llvm/bin/ld.lld -Wl,-no-as-needed -Wl,-z,relro,-z,now -lstdc++ -lm -lcrypto -lcurl -ldl -ljpeg -llzma -lpthread -lsqlite3 -lssl -ltiff -lwebp -lz -lzstd -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/bz2_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/fftw_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/z_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/png_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/open_ssl_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/curl_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/sqlite_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/jpeg_turbo_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/lzma_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/gif_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/webp_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/zstd_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/tiff_lib/lib -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/proj_lib/lib -Wl,--export-dynamic -Wl,-rpath-link,/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/lib -o /root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/bin/r.mapcalc OBJ.x86_64-pc-linux-gnu/column_shift.o OBJ.x86_64-pc-linux-gnu/evaluate.o OBJ.x86_64-pc-linux-gnu/expression.o OBJ.x86_64-pc-linux-gnu/function.o OBJ.x86_64-pc-linux-gnu/main.o OBJ.x86_64-pc-linux-gnu/map.o OBJ.x86_64-pc-linux-gnu/xarea.o OBJ.x86_64-pc-linux-gnu/xcoor.o OBJ.x86_64-pc-linux-gnu/xres.o OBJ.x86_64-pc-linux-gnu/xrowcol.o OBJ.x86_64-pc-linux-gnu/mapcalc.yy.o OBJ.x86_64-pc-linux-gnu/mapcalc.tab.o    -lgrass_calc.7.8 -lgrass_raster.7.8 -lgrass_gis.7.8 -lgrass_datetime.7.8   -lz   -lbz2   -lzstd         -lgrass_gproj.7.8 -lgrass_gis.7.8 -lgrass_datetime.7.8   -lz   -lbz2   -lzstd          -lproj    -lgrass_gis.7.8 -lgrass_datetime.7.8   -lz   -lbz2   -lzstd          -lgrass_gis.7.8 -lgrass_datetime.7.8   -lz   -lbz2   -lzstd         -lgrass_raster.7.8 -lgrass_gis.7.8 -lgrass_datetime.7.8   -lz   -lbz2   -lzstd         -lgrass_gproj.7.8 -lgrass_gis.7.8 -lgrass_datetime.7.8   -lz   -lbz2   -lzstd          -lproj    -lgrass_btree.7.8
```
I suspect something is wrong with the linker flags or the flag order, but maybe this is going to be immediately obvious to someone so posting it here.

**To Reproduce**

Currently the easiest way to repro is to build the whole thing, including dependencies, inside a Docker container:

```
git clone git@github.com:attilaolah/wasm
cd wasm
docker build -t wasm docker
docker run -it -v "${PWD}:/build" wasm
```

Then once inside the container, build like so:

```
bazel build //lib/grass
```

It will download the same compiler and everything, build dependencies first and then attempt to build grass itself.

**Expected behavior**

It should build.

**System description (please complete the following information):**

- Operating System: Ubuntu 20.10 from the Docker hub
- GRASS GIS version 7.8.5, haven't tried other versions so far.

**Comment by nilason** [comment]
Could it be that `--export-dynamic` trigger this?

**Comment by attilaolah** [comment]
Hmm, could be. I'm not sure what is adding that flag though, the default linker flags that I'm passing are [set up](https://github.com/attilaolah/wasm/commit/080bb9a0436c536c8d6facf199aa0682d24c36b6#diff-da000eb89e752fa1be1e4b3059bf4497940ada0cbb691cd18f76ed2a2f6e733cR93-R100) to as so:

```starlark
link_flags = [
    "-B{}/bin".format(LLVM_PATH),
    "-fuse-ld={}/bin/ld.lld".format(LLVM_PATH),
    "-Wl,-no-as-needed",
    "-Wl,-z,relro,-z,now",
    "-lstdc++",
    "-lm",
],
```

And this is my full configure line:

```
ARFLAGS="rcsD" ASFLAGS="-U_FORTIFY_SOURCE -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -fstack-protector -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__='\"redacted\"' -D__TIME__='\"redacted\"' -D__TIMESTAMP__='\"redacted\"'" CFLAGS="-U_FORTIFY_SOURCE -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -fstack-protector -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__='\"redacted\"' -D__TIME__='\"redacted\"' -D__TIMESTAMP__='\"redacted\"'" CXXFLAGS="-U_FORTIFY_SOURCE -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -fstack-protector -std=c++17 -no-canonical-prefixes -Wno-builtin-macro-redefined -D__DATE__='\"redacted\"' -D__TIME__='\"redacted\"' -D__TIMESTAMP__='\"redacted\"'" LDFLAGS="-B${EXT_BUILD_ROOT}/external/llvm/bin -fuse-ld=${EXT_BUILD_ROOT}/external/llvm/bin/ld.lld -Wl,-no-as-needed -Wl,-z,relro,-z,now -lstdc++ -lm -lcrypto -lcurl -ldl -ljpeg -llzma -lpthread -lsqlite3 -lssl -ltiff -lwebp -lz -lzstd -L$EXT_BUILD_DEPS/bz2_lib/lib -L$EXT_BUILD_DEPS/fftw_lib/lib -L$EXT_BUILD_DEPS/z_lib/lib -L$EXT_BUILD_DEPS/png_lib/lib -L$EXT_BUILD_DEPS/open_ssl_lib/lib -L$EXT_BUILD_DEPS/curl_lib/lib -L$EXT_BUILD_DEPS/sqlite_lib/lib -L$EXT_BUILD_DEPS/jpeg_turbo_lib/lib -L$EXT_BUILD_DEPS/lzma_lib/lib -L$EXT_BUILD_DEPS/gif_lib/lib -L$EXT_BUILD_DEPS/webp_lib/lib -L$EXT_BUILD_DEPS/zstd_lib/lib -L$EXT_BUILD_DEPS/tiff_lib/lib -L$EXT_BUILD_DEPS/proj_lib/lib" AR="/.${EXT_BUILD_ROOT}/external/llvm/bin/llvm-ar" CC="/.${EXT_BUILD_ROOT}/external/llvm/bin/clang" CXX="/.${EXT_BUILD_ROOT}/external/llvm/bin/clang" CPPFLAGS="-I$EXT_BUILD_DEPS/bz2_lib/include -I$EXT_BUILD_DEPS/fftw_lib/include -I$EXT_BUILD_DEPS/z_lib/include -I$EXT_BUILD_DEPS/png_lib/include -I$EXT_BUILD_DEPS/open_ssl_lib/include -I$EXT_BUILD_DEPS/curl_lib/include -I$EXT_BUILD_DEPS/sqlite_lib/include -I$EXT_BUILD_DEPS/jpeg_turbo_lib/include -I$EXT_BUILD_DEPS/lzma_lib/include -I$EXT_BUILD_DEPS/gif_lib/include -I$EXT_BUILD_DEPS/webp_lib/include -I$EXT_BUILD_DEPS/zstd_lib/include -I$EXT_BUILD_DEPS/tiff_lib/include -I$EXT_BUILD_DEPS/proj_lib/include" "$BUILD_TMPDIR/configure" --prefix=$BUILD_TMPDIR/$INSTALL_PREFIX --enable-shared=no --with-bzlib --with-cxx --with-fftw --with-nls --with-png --with-pthread --with-regex --with-sqlite --with-tiff --with-zstd --without-blas --without-cairo --without-freetype --without-gdal --without-lapack --without-mysql --without-odbc --without-opencl --without-opendwg --without-opengl --without-openmp --without-postgres --without-readline
```

No `--export-dynamic` in there. I'll attach my entire [config.log](https://github.com/OSGeo/grass/files/5977421/config.log) in case that helps… But will try to disable that `--export-dynamic` flag somehow.

Thatks for the tip.

**Comment by nilason** [comment]
Please try compile with `-fno-common`.  See https://reviews.llvm.org/D75056

**Comment by attilaolah** [comment]
I'll try, but I'm building with Clang 11.0.1, and I believe it already has `-fno-common` as the default. Additionally, I don't think building with `-fno-common` would result in this error to go away: if anything, it should cause this error.

I'll also try to build with `-fcommon`, if that succeeds, I'll send a PR to add the corresponding `extern`s.

**Comment by nilason** [comment]
Building with Clang on Mac causes no problems and I can't reproduce this (with and without -f(no)common).

Out of curiosity, would this particular problem remain if you change the following lines:

https://github.com/OSGeo/grass/blob/f596255917f974cafec921677037b3e228ce69f0/lib/calc/calc.c#L12

https://github.com/OSGeo/grass/blob/f596255917f974cafec921677037b3e228ce69f0/raster/r.mapcalc/map.c#L26


to `static int columns; `

**Comment by attilaolah** [comment]
I'll check; I'm trying to repro this now, I modified my environment somehow and I'm getting different build problems :(

But thanks for the tip, I'll try again if I encounter this. In the meantime feel free to close this as _not reproducible_.

**Comment by attilaolah** [comment]
Oh wait, I managed to repro this, after fixing a couple other problems. Let me try your suggestion.

**Comment by attilaolah** [comment]
OK, just changing those two declarations to `static` is not enough:

```
map.c:27:12: error: static declaration of 'columns' follows non-static declaration
static int columns;
           ^
./globals.h:11:26: note: previous declaration is here
extern int depths, rows, columns;
```

I also tried building with `-fcommon` or `-fno-common`, and neither helps. I also tried removing `-Wl,--export-dynamic`, that didn't help either :(

**Comment by attilaolah** [comment]
One thing that helps though, is removing the `int columns; line entirely from `raster/r.mapcalc/map.c`:

https://github.com/OSGeo/grass/blob/f596255917f974cafec921677037b3e228ce69f0/raster/r.mapcalc/map.c#L26

It doesn't seem to be required anyway, since the file includes `"globals.h"`, which contains the extern declaration.

Would it be OK to remove that line? (And see if CI passes at least…)

**Comment by nilason** [comment]
That re-definition of columns seems to me to be indeed unnecessary.
Compiles for me without it and quick test of r.mapcalc didn't produce any issue.

**Comment by attilaolah** [comment]
I hit another case of this error now, in `raster/r.buildvrt/main.c`, which redefines `cmp_wnd` from `lib/raster/vrt.c`:

https://github.com/OSGeo/grass/blob/7da47c1c0cfa5df5e00d55772386ef75ae00c16e/raster/r.buildvrt/main.c#L30-L54

https://github.com/OSGeo/grass/blob/7da47c1c0cfa5df5e00d55772386ef75ae00c16e/lib/raster/vrt.c#L21-L45

However, this time removing the definition from `main.c` is not enough, as the function doesn't seem to be declared in any of the header files…

**Comment by attilaolah** [comment]
I guess renaming the `cmp_wnd` in the `main.c` file would solve that second issue.

**Comment by nilason** [comment]
The question remains, what causes these for your setup?
What about -Wthread-safety ?

**Comment by attilaolah** [comment]
Lemme try.

**Comment by attilaolah** [comment]
Removing the flag `-fuse-ld=/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/352/execroot/wasm/external/llvm/bin/ld.lld` to use the default system linker, `/usr/bin/ld`, changes the error message to a different one, but still essentially the same:

```
/usr/bin/ld: /root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/lib/libgrass_raster.7.8.a(vrt.o): in function `cmp_wnd':
vrt.c:(.text+0x0): multiple definition of `cmp_wnd'; OBJ.x86_64-pc-linux-gnu/main.o:main.c:(.text+0x0): first defined here
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by attilaolah** [comment]
I now retried with [LLVM+Clang 12.0.0-rc1](https://github.com/llvm/llvm-project/releases/tag/llvmorg-12.0.0-rc1), still the same error.

Here is the complete build command, line-by-line, I also removed some flags that I thought could be the problem, but I still get the same error.

```sh
/build/clang+llvm-12.0.0-rc1-x86_64-linux-gnu-ubuntu-20.10/bin/clang \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/lib \
  -B/build/clang+llvm-12.0.0-rc1-x86_64-linux-gnu-ubuntu-20.10/bin \
  -fuse-ld=/build/clang+llvm-12.0.0-rc1-x86_64-linux-gnu-ubuntu-20.10/bin/ld.lld \
  -lstdc++ \
  -lm \
  -lcrypto \
  -lcurl \
  -ldl \
  -ljpeg \
  -llzma \
  -lpthread \
  -lsqlite3 \
  -lssl \
  -ltiff \
  -lwebp \
  -lz \
  -lzstd \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/bz2_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/fftw_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/z_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/png_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/open_ssl_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/curl_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/sqlite_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/jpeg_turbo_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/lzma_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/gif_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/webp_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/zstd_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/tiff_lib/lib \
  -L/root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.ext_build_deps/proj_lib/lib \
  -lgrass_raster.7.8 \
  -lgrass_gis.7.8 \
  -lgrass_datetime.7.8 \
  -lz \
  -lbz2 \
  -lzstd \
  -lgrass_gproj.7.8 \
  -lgrass_gis.7.8 \
  -lgrass_datetime.7.8 \
  -lz \
  -lbz2 \
  -lzstd \
  -lproj \
  -lgrass_gis.7.8 \
  -lgrass_datetime.7.8 \
  -lz \
  -lbz2 \
  -lzstd \
  OBJ.x86_64-pc-linux-gnu/link.o \
  OBJ.x86_64-pc-linux-gnu/main.o \
  -o /root/.cache/bazel/_bazel_root/7b7747ec045ae606eb720a1222f56098/sandbox/processwrapper-sandbox/9/execroot/wasm/bazel-out/k8-fastbuild/bin/lib/grass/grass_lib.build_tmpdir/dist.x86_64-pc-linux-gnu/bin/r.buildvrt
```

**Comment by nilason** [comment]
> I guess renaming the `cmp_wnd` in the `main.c` file would solve that second issue.

The two cpp_wnd are practically identical, but both are supposed to have file scope as I see it. If you state those two `static` as we tried before... (`static int cmp_wnd(const void *a, const void *b)`)...

**Comment by attilaolah** [comment]
OK let me try that. I'm not familiar with C scoping semantics, in C++ you would put these in an anonymous namespace (or some other namespace) to solve these issues. Again pardon my rusty C, I'll try the `static` definition in a moment.

**Comment by attilaolah** [comment]
Hmm, looks like changing just the function in the `main.c` to `static int cmp_wnd(…)` is enough to fix this particular problem.

**Comment by attilaolah** [comment]
I really wish I knew where these errors are coming from. Maybe I'll have some time to bisect this with different versions of Clang.

**Comment by metzm** [comment]
@attilaolah This is the first time these errors occur. Regarding the first post, `columns` in `lib/calc.c` and in `raster/r.mapcalc/map.c` are two completely independent variables in different scopes. The same variable names, also with some `extern ..` declaration in a module-specific (!) header, are used all over the place. If you get this `duplicate symbol` error with `r.mapcalc`, you should also get this error with many other GRASS libs and modules. Nobody else gets these errors.

**Comment by attilaolah** [comment]
OK, thanks. I guess I'm either using the wrong compiler, or the wrong combination of linker flags. Feel free to ignore my pull requests then.

For now I'll probably patch these on my side at build time so that I could move forward (patching seems harmless for my use case), when I find some time I'll try to root cause what causes the linker errors.

Thanks everyone for looking into these for me.

**Comment by nilason** [comment]
I did experiment building with clang on Linux:

https://github.com/nilason/grass/actions/runs/575040344

with mixed results. See above referenced builds for ideas. 
`-std=gnu11 -fexceptions -fno-common  -fPIC ` seemed to work.
Please report back if you find the cause.

**Comment by attilaolah** [comment]
Oh good, I can try those flags. Currently I'm not using `-fPIC`, but then I'm also only interested in building the static libraries (I don't care about the binaries, but it's easier if I can get the whole build to pass).

I also used `-std=c++17`, so maybe `-std=gnu11` would work. I don't have `-fno-common` but then that should be the default.

I'll let you know if I can get this to build.

So far I am down to a few minor issues:

- The duplicate definitions in the two cases mentioned in this bug.
- Building after a `configure --enable-shared=no` still tries to build `lib/python/ctypes` but fails, since those need shared libs. A simple `sed` to remove that directory from the `Makefile` seems to fix the issue.
- #1376
- #1377
- I can't seem to get the build to find my PROJ `proj.db` file, even after setting `--with-proj-share` to various values.

Apart from these things above, the Clang build seems to be fine (static libraries build with no problem).

**Comment by nilason** [comment]
There are also some docker files: https://github.com/OSGeo/grass/tree/master/docker, which could give some hints.

**Comment by attilaolah** [comment]
In fact, if there was a configure option to skip building all the binaries and Python bindings, and just build the static libs, that would be good enough for me — but while I'm at it, I might as well try and build the whole thing.

And thanks for the Docker files, will give those a try.

**Comment by nilason** [comment]
There is a `make libs` alternative if I’m not mistaken.

**Comment by attilaolah** [comment]
Ah right! I first thought that `make install` would depend on the default target, but apparently no, I can just do `make libs` and `make install`.

Thanks, that's probably enough for me then. I don't need the binaries for now.

**Comment by HuidaeCho** [comment]
> `columns` in `lib/calc.c` and in `raster/r.mapcalc/map.c` are two completely independent variables in different scopes.

Agreed. I don't think it's a good idea to merge  #1380, which eliminates the `r.mapcalc` version of `columns`.

**Comment by nilason** [comment]
@attilaolah As I can't reproduce this (yet) I need to ask: after the `columns` and the `cmp_wnd ` issues, were there more of the same kind?

**Comment by marisn** [comment]
It took me many hours of rereading same [SO answers](https://stackoverflow.com/questions/1433204/how-do-i-use-extern-to-share-variables-between-source-files) to get to a potential understanding of this problem (yes, I am not good at C).

This bug is easy to reproduce – just `./configure` with `--disable-shared` and you'll get it too.

Facts:
* `include/calc.h` has a declaration of `columns`
* `lib/calc/calc.c` has a definition (and initialization) of `columns`
* `raster/r.mapcalc/global.h` has a declaration of `columns`
* `raster/r.mapcalc/map.c` has a definition (and initialization) of `columns`

Per [ISO/IEC 9899:201x 6.9 External definitions](http://port70.net/%7Ensz/c/c11/n1570.html#6.9): "somewhere in the entire program there shall be exactly one external definition for the identifier". Thus lib/calc and r.mapcalc are fine as long as they are not linked into a single program. As lib/calc is just an exposure of r.mapcalc to whole GRASS, this is a no-go (they must link into a single program). This case is covered in [ISO/IEC 9899:201x J.2 Undefined behavior](http://port70.net/%7Ensz/c/c11/n1570.html#J.2): "An identifier with external linkage is used, but in the program there does not exist exactly one external definition for the identifier, or the identifier is not used and there exist multiple external definitions for the identifier". Still we also have to take into account [ISO/IEC 9899:201x J.5.11 Multiple external definitions](http://port70.net/~nsz/c/c11/n1570.html#J.5.11): "There may be more than one external definition for the identifier of an object, with or without the explicit use of the keyword extern; if the definitions disagree, or more than one is initialized, the behavior is undefined". This last bit explains why PR #1380 solves the problem – it removes second implicit initialization ([ISO/IEC 9899:201x 6.2.4 Storage durations of objects](http://port70.net/~nsz/c/c11/n1570.html#6.2.4p3) and [ISO/IEC 9899:201x 6.7.9 Initialization](http://port70.net/~nsz/c/c11/n1570.html#6.7.9p10)) thus pushing the issue under a rug.

TL;DR:
lib/calc/ has a bad design and should be reworked to not use a global variable as it makes its usage error-prone.

**Comment by attilaolah** [comment]
> @attilaolah As I can't reproduce this (yet) I need to ask: after the `columns` and the `cmp_wnd ` issues, were there more of the same kind?

No, only these two cases: everything else links just fine.

And thanks @marisn for the detailed explanation.

**Comment by nilason** [comment]
> It took me many hours of rereading same [SO answers](https://stackoverflow.com/questions/1433204/how-do-i-use-extern-to-share-variables-between-source-files) to get to a potential understanding of this problem (yes, I am not good at C).
> 
> This bug is easy to reproduce – just `./configure` with `--disable-shared` and you'll get it too.

Terrific, thanks for the great research! This will provide good basis to fix the issues properly.

In the end I started to suspect -shared/--disable-shared was involved but never pursued it.


> TL;DR:
> lib/calc/ has a bad design and should be reworked to not use a global variable as it makes its usage error-prone.

Agreed.

**Comment by metzm** [comment]
The historical explanation for this definition and declaration of `columns` is that `lib/calc` was previously a part of `r.mapcalc` and has then moved out to a separate lib. This does not excuse the missing adjustments.
